helix-core/src/main/java/org/apache/helix/util/HelixUtil.java

```diff
@@ -519,13 +519,20 @@ public static InstanceConfig composeInstanceConfig(String instanceName) {
   }
 
   /**
-   * Checks whether or not the cluster is in management mode.
+   * Checks whether or not the cluster is in management mode. It checks pause signal
+   * and live instances: whether any live instance is not in normal status, eg. frozen.
    *
-   * @param cache
-   * @return
+   * @param pauseSignal pause signal
+   * @param liveInstanceMap map of live instances
+   * @param enabledLiveInstances set of enabled live instance names. They should be all included
+   *                             in the liveInstanceMap.
+   * @return true if cluster is in management mode; otherwise, false
    */
-  public static boolean inManagementMode(BaseControllerDataProvider cache) {
-    // TODO: implement the logic. Parameters can also change
-    return true;
+  public static boolean inManagementMode(PauseSignal pauseSignal,
+      Map<String, LiveInstance> liveInstanceMap, Set<String> enabledLiveInstances) {
+    // Check pause signal and abnormal live instances (eg. in freeze mode)
+    // TODO: should check maintenance signal when moving maintenance to management pipeline
+    return pauseSignal != null || enabledLiveInstances.stream()
+        .anyMatch(instance -> liveInstanceMap.get(instance).getStatus() != null);
```

Comment:
IMO, getStatus() != null is little bit hacky. Better to check if the status is PAUSED or whatever other specific status.